Segment/Classes/SEGWebhookIntegration.m

```diff
@@ -0,0 +1,208 @@
+        if (code == 429) {
+            // 429 response codes. Retry.
+            SEGLog(@"Server limited client with response code %d.", code);
+            return;
+        }
+        if (code < 500) {
+            // non-429 4xx response codes. Don't retry.
+            SEGLog(@"Server rejected payload with HTTP code %d.", code);
+            return;
+        }
+
+        // 5xx response codes. Retry.
+        SEGLog(@"Server error with HTTP code %d.", code);
+    }];
+    [task resume];
+}
+
+// Merges user provided integration options with bundled integrations.
+- (NSDictionary *)integrationsDictionary:(NSDictionary *)integrations
+{
+    NSMutableDictionary *dict = [integrations ?: @{} mutableCopy];
+    for (NSString *integration in self.analytics.bundledIntegrations) {
+        // Don't record Segment.io in the dictionary. It is always enabled.
+        if ([integration isEqualToString:@"Segment.io"]) {
+            continue;
+        }
+        dict[integration] = @NO;
+    }
+    return [dict copy];
+}
+
+// Code borrowed from SEGSegmentIntegration.enqueueAction
+- (void)enqueue:(NSString *) type dictionary:(NSMutableDictionary *) payload context:(NSDictionary *)context integrations:(NSDictionary *)integrations
+{
+    payload[@"type"] = type;
+    if (![type isEqualToString:@"alias"]) {
+        [payload setValue:[SEGState sharedInstance].userInfo.userId forKey:@"userId"];
+    }
+    [payload setValue:[self.analytics getAnonymousId] forKey:@"anonymousId"];
+    [payload setValue:[self integrationsDictionary:integrations] forKey:@"integrations"];
+    [payload setValue:[context copy] forKey:@"context"];
+
+    [self dispatchBackground:^{
+        SEGLog(@"%@ Enqueueing payload %@ through %@", self, type, self.name);
+        NSDictionary *queuePayload = [payload copy];
+        [self sendPayloadToWebhook:queuePayload];
+    }];
+}
+
+- (void)dispatchBackground:(void (^)(void))block
+{
+    seg_dispatch_specific_async(_serialQueue, block);
+}
+
+- (NSString *)userId
+{
+    return [SEGState sharedInstance].userInfo.userId;
+}
+
+- (void)identify:(SEGIdentifyPayload *)payload
+{
+    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
+    [dictionary setValue:payload.traits forKey:@"traits"];
+    [dictionary setValue:payload.timestamp forKey:@"timestamp"];
+    [dictionary setValue:payload.messageId forKey:@"messageId"];
+
+    [self enqueue:@"identify" dictionary:dictionary context:payload.context integrations:payload.integrations];
+}
+
+- (void)track:(SEGTrackPayload *)payload
+{
+    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
+    [dictionary setValue:payload.event forKey:@"event"];
+    [dictionary setValue:payload.properties forKey:@"properties"];
+    [dictionary setValue:payload.timestamp forKey:@"timestamp"];
+    [dictionary setValue:payload.messageId forKey:@"messageId"];
+
+    [self enqueue:@"track" dictionary:dictionary context:payload.context integrations:payload.integrations];
```

Comment:
When working with NSMutableDictionary it's usually best to convert it back to a regular dictionary before sending it off so other's can't modify it. Unless that is the intended action.

Reply:
this is intended